File: services/academic-service/src/test/java/com/visor/school/academic/service/TeacherServiceTest.java
package com.visor.school.academic.service;

import com.visor.school.academic.model.EmploymentStatus;
import com.visor.school.academic.model.Teacher;
import com.visor.school.academic.repository.TeacherRepository;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class TeacherServiceTest {

    @Mock
    private TeacherRepository teacherRepository;

    @Mock
    private EmployeeIdGenerator employeeIdGenerator;

    @InjectMocks
    private TeacherService teacherService;

    private final Teacher testTeacher = new Teacher(
            "T54321",
            UUID.randomUUID(),
            List.of("PhD in Physics"),
            List.of("Physics", "Quantum Mechanics"),
            LocalDate.of(2018, 8, 15),
            "Science",
            EmploymentStatus.ACTIVE
    );

    @Test
    void createTeacherShouldSaveAndReturnTeacher() {
        UUID userId = testTeacher.getUserId();
        Teacher savedTeacher = new Teacher(
                "T54321",
                userId,
                testTeacher.getQualifications(),
                testTeacher.getSubjectSpecializations(),
                testTeacher.getHireDate(),
                testTeacher.getDepartment(),
                testTeacher.getEmploymentStatus()
        );
        savedTeacher.setId(UUID.randomUUID());

        when(employeeIdGenerator.generateEmployeeId()).thenReturn("T54321");
        when(teacherRepository.findByUserId(any())).thenReturn(Optional.empty());
        when(teacherRepository.save(any(Teacher.class))).thenReturn(savedTeacher);

        Teacher createdTeacher = teacherService.createTeacher(
                userId,
                testTeacher.getQualifications(),
                testTeacher.getSubjectSpecializations(),
                testTeacher.getHireDate(),
                testTeacher.getDepartment()
        );

        assertNotNull(createdTeacher.getId());
        assertNotNull(createdTeacher.getEmployeeId());
        verify(teacherRepository).save(any(Teacher.class));
    }

    @Test
    void getTeacherByIdShouldReturnTeacher() {
        UUID teacherId = UUID.randomUUID();
        when(teacherRepository.findById(teacherId)).thenReturn(Optional.of(testTeacher));

        Teacher foundTeacher = teacherService.getTeacherById(teacherId);

        assertEquals(testTeacher, foundTeacher);
        verify(teacherRepository).findById(teacherId);
    }

    @Test
    void getTeachersByStatusShouldReturnAList() {
        when(teacherRepository.findByEmploymentStatus(EmploymentStatus.ACTIVE))
                .thenReturn(List.of(testTeacher));

        List<Teacher> teachers = teacherService.getTeachersByStatus(EmploymentStatus.ACTIVE);

        assertFalse(teachers.isEmpty());
        assertEquals(1, teachers.size());
        verify(teacherRepository).findByEmploymentStatus(EmploymentStatus.ACTIVE);
    }

    @Test
    void getTeachersByDepartmentShouldReturnAList() {
        when(teacherRepository.findByDepartment("Science"))
                .thenReturn(List.of(testTeacher));

        List<Teacher> teachers = teacherService.getTeachersByDepartment("Science");

        assertFalse(teachers.isEmpty());
        assertEquals(1, teachers.size());
        verify(teacherRepository).findByDepartment("Science");
    }
}

File: services/academic-service/src/test/kotlin/com/visor/school/academicservice/service/TeacherServiceTest.kt
package com.visor.school.academicservice.service

import com.visor.school.academicservice.model.EmploymentStatus
import com.visor.school.academicservice.model.Teacher
import com.visor.school.academicservice.repository.TeacherRepository
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.*
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.PageRequest
import java.time.LocalDate
import java.util.Optional
import java.util.UUID

@ExtendWith(MockitoExtension::class)
class TeacherServiceTest {

    @Mock
    private lateinit var teacherRepository: TeacherRepository

    @Mock
    private lateinit var employeeIdGenerator: EmployeeIdGenerator

    @InjectMocks
    private lateinit var teacherService: TeacherService

    private val testTeacher = Teacher(
        userId = UUID.randomUUID(),
        employeeId = "T54321",
        qualifications = listOf("PhD in Physics"),
        subjectSpecializations = listOf("Physics", "Quantum Mechanics"),
        hireDate = LocalDate.of(2018, 8, 15),
        employmentStatus = EmploymentStatus.ACTIVE,
        department = "Science"
    )

    @Test
    fun `create teacher should save and return teacher`() {
        // Given
        val savedTeacher = Teacher(
            id = UUID.randomUUID(),
            userId = testTeacher.userId,
            employeeId = "T54321",
            qualifications = testTeacher.qualifications,
            subjectSpecializations = testTeacher.subjectSpecializations,
            hireDate = testTeacher.hireDate,
            employmentStatus = testTeacher.employmentStatus,
            department = testTeacher.department
        )
        whenever(employeeIdGenerator.generateEmployeeId()).thenReturn("T54321")
        whenever(teacherRepository.findByUserId(any())).thenReturn(Optional.empty())
        whenever(teacherRepository.save(any<Teacher>())).thenReturn(savedTeacher)

        // When
        val createdTeacher = teacherService.createTeacher(
            testTeacher.userId,
            testTeacher.qualifications,
            testTeacher.subjectSpecializations,
            testTeacher.hireDate,
            testTeacher.department
        )

        // Then
        assertNotNull(createdTeacher.id)
        assertNotNull(createdTeacher.employeeId)
        verify(teacherRepository).save(any<Teacher>())
    }

    @Test
    fun `get teacher by id should return teacher`() {
        // Given
        val teacherId = UUID.randomUUID()
        whenever(teacherRepository.findById(teacherId)).thenReturn(Optional.of(testTeacher))

        // When
        val foundTeacher = teacherService.getTeacherById(teacherId)

        // Then
        assertEquals(testTeacher, foundTeacher)
        verify(teacherRepository).findById(teacherId)
    }

    @Test
    fun `get teachers by status should return a list`() {
        // Given
        whenever(teacherRepository.findByEmploymentStatus(EmploymentStatus.ACTIVE))
            .thenReturn(listOf(testTeacher))

        // When
        val teachers = teacherService.getTeachersByStatus(EmploymentStatus.ACTIVE)

        // Then
        assertFalse(teachers.isEmpty())
        assertEquals(1, teachers.size)
        verify(teacherRepository).findByEmploymentStatus(EmploymentStatus.ACTIVE)
    }

    @Test
    fun `get teachers by department should return a list`() {
        // Given
        whenever(teacherRepository.findByDepartment("Science"))
            .thenReturn(listOf(testTeacher))

        // When
        val teachers = teacherService.getTeachersByDepartment("Science")

        // Then
        assertFalse(teachers.isEmpty())
        assertEquals(1, teachers.size)
        verify(teacherRepository).findByDepartment("Science")
    }
}
